HW2/P2/parallel_vector.pyx

```diff
@@ -81,10 +80,15 @@ cpdef move_data_fine_grained(np.int32_t[:] counts,
    ##########
    with nogil:
        for r in range(repeat):
-           for idx in range(src.shape[0]):
+           for idx in prange(src.shape[0], num_threads=4):  
                if counts[src[idx]] > 0:
+                   # Add locks and release when done
+                   acquire(&locks[dest[idx]])
                    counts[dest[idx]] += 1
+                   release(&locks[dest[idx]])
+                   acquire(&locks[src[idx]])
                    counts[src[idx]] -= 1
+                   release(&locks[src[idx]])
```

Comment:
This procedure definitely prevents deadlock and double-locking. However, counts[src[idx]] may change between the time you check its value and the time you decrement it because other threads can edit it in the meantime since it's not locked yet. Therefore, it is possible that this code will decrement counts[src[idx]] to a negative value, which is incorrect behavior. To prevent this, you would need to lock src[idx] before checking its count.

I suspect you should lock both src and dest prior to modifying either, but I don't have an explanation off the top of my head so I could be wrong.